### PR TITLE
Fix log splitting issue in FilterEvents

### LINKED ISSUE #12535: Enable FilterEvents to read splitters from a TableWorkspace
This issue was originally [TRAC 11697](http://trac.mantidproject.org/mantid/ticket/11697)

Currently FilterEvents takes only SplittersWorkspace and MatrixWorkspace for splitters.  Users may want to generate splitters via python script in table workspace.  Enable FilterEvents to support this format.

There is a bug that has been reported that if the unit of workspace is not of time-of-flight, there would be no exception raised.

### LINKED ISSUE #16756: Fix reported issues with FilterEvents
There are some relatively minor issues with FilterEvents.
1. Slow sample log will be split with MatrixWorkspace splitter;
2. If the last splitter's stop time is left non-physically large, such as 3153600000 seconds (100 years X 356 days X 24 hours X 3600 seconds;

Also the codes should be refactored such that the underlying codes for MatrixWorkspace splitter and SplittersWorokspace splitter should be same.

### Expected behavior
1. Any TimeSeriesProperty in Run object shall be only split once, and the split time series property shall we stored to a map of TimeSeriesProperties;
2. Method TimeSeriesProperty.filterByTimes() will be replaced (but not removed) by TimeSeriesProperty.splitByTimeVectors(vec_times, vec_targets) in FilterEvents.  splitByTimeVectors() shall consider the fast log situation;
3.

### PR DESCRIPTION
**Pull request by wdzhou**
Description of work.

**To test:**

1. Find an EventNexus file to split from.
2. Test using TableWorkspace for splitters
  - Refer to the doc test of **FilterEvents** to create a TableWorkspace
  - Run *FilterEvents* with this TableWorkspace
3. Test splitting sample logs for splitters given by MatrixWorkspace and TableWorkspace
  - Check the sample logs of the output workspaces from step (2). 
  - Run **GenerateEventsFilter** to create MatrixWorkspace splitters by set 'FastLog'  to True.
  - Run **FilterEvents** with the MatrixWorkspace splitterd and check sample logs to be properly split
4. Check whether all the output workspaces has a sample log named *splitter*, which reflects the splitter that is applied to it.

Fixes #16756.
Fixes #12535.

*The change is reported in the FrameWork section in release note*

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by wdzhou** [comment]
** Road Map**

1. Complete all the unit test;
2. Try to replace linear search in splitting by binary search;
3. Speed test on large log (e.g. proton charge) with relatively high frequency splitters;
4. Get data from Garrett for his pulse magnetic data;

**Review by peterfpeterson — CHANGES_REQUESTED** [review]
Looks good other than changing some log levels

File: Framework/Algorithms/src/FilterEvents.cpp [hunk 1]
```diff
@@ -1000,6 +1581,144 @@ void FilterEvents::splitLog(EventWorkspace_sptr eventws, std::string logname,
...
+}
+
+/** add the splitter TimeSeriesProperty logs to each workspace
+ * @brief FilterEvents::mapSplitterTSPtoWorkspaces
+ * @param split_tsp_vec
+ */
+void FilterEvents::mapSplitterTSPtoWorkspaces(
+    const std::vector<Kernel::TimeSeriesProperty<int> *> &split_tsp_vec) {
+  if (m_useSplittersWorkspace) {
+    g_log.warning() << "There are " << split_tsp_vec.size()
```
Comment (wdzhou) [inline Framework/Algorithms/src/FilterEvents.cpp]: Fixed in the last commit.

File: Framework/Algorithms/src/FilterEvents.cpp [hunk 2]
```diff
@@ -1000,6 +1581,144 @@ void FilterEvents::splitLog(EventWorkspace_sptr eventws, std::string logname,
...
+                    << " TimeSeriesPropeties.\n";
+    std::map<int, DataObjects::EventWorkspace_sptr>::iterator miter;
+    for (miter = m_outputWorkspacesMap.begin();
+         miter != m_outputWorkspacesMap.end(); ++miter) {
+      g_log.warning() << "Output workspace index: " << miter->first << "\n";
```
Comment (wdzhou) [inline Framework/Algorithms/src/FilterEvents.cpp]: Fixed in the last commit.

**Review by peterfpeterson — APPROVED** [review]

**Comment by peterfpeterson** [comment]
:shipit:

**Pull request merged**